### DISCUSSION TITLE
C# wrappers of Existing c++ GDExtension - 2024 Update

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [godotengine/godot-proposals#7024] [Scripting]
### 12th April 2024 (FINALLY)

@Delsin-Yu

https://github.com/Delsin-Yu/CSharp-Wrapper-Generator-for-GDExtension

```cs
        // This construct an instance of the underlying 
        // GDExtension Node, we are using a static method
        // here is we need to do manual black magic for
        // setting up the wrapper script and the underlying 
        // GDExtension node instance.
        var joltGeneric6DofJoint3D = JoltGeneric6DOFJoint3D.Construct();
        AddChild(joltGeneric6DofJoint3D);
        GD.Print(joltGeneric6DofJoint3D.AngularLimitYEnabled);
        joltGeneric6DofJoint3D.AngularLimitYEnabled = false;
        GD.Print(joltGeneric6DofJoint3D.AngularLimitYEnabled);
        joltGeneric6DofJoint3D.AngularLimitYEnabled = true;
        GD.Print(joltGeneric6DofJoint3D.AngularLimitYEnabled);
```

<details>
<summary>5th April 2024 with Official discussion from the Godot Team </summary>

# Official discussion from the Godot Team
[Godot C# GDExtension Roadmap Discussion](https://github.com/godotengine/godot-proposals/issues/7895)

### 5th April 2024 
# The Official [Godot4 GDExtension](https://github.com/godotengine/godot-proposals/issues/7895) for .NET is Here
https://github.com/godotengine/godot/pull/72883#issuecomment-2040889306

Main proposal: https://github.com/godotengine/godot-proposals/issues/7895

Community discussion: https://chat.godotengine.org/channel/dotnet?msg=Kyvjpbutadr2bnzvG

</details>

<details>
<summary>25th Jan 2024 Update</summary>

## [Proposed and prototyped](https://github.com/godotengine/godot-proposals/issues/8191#issuecomment-1910977324) by @DmitriySalnikov

![image](https://github.com/godotengine/godot-proposals/assets/49812372/9d845b19-b0c8-4e6c-be14-412a6b2a75e5)
</details>

<details>
<summary>28th Nov 2023 Update</summary>

# 28th Nov 2023 status
## [Auto generate C# bindings for gdextensions](https://github.com/godotengine/godot-proposals/issues/8191#issuecomment-1830490771)
</details>
<details>
<summary>13th Sept 2023 Update</summary>
# 18th Oct 2023 Update (After Godot 4.2 Beta 1)
https://github.com/godotengine/godot/pull/82331#issuecomment-1767267091

# 13th Sept 2023 Update
A menu item of Generate C# Bindings has been added
https://github.com/DmitriySalnikov/godot_debug_draw_3d

![notepad++_BItVY8zquj](https://github.com/godotengine/godot-proposals/assets/49812372/2c7d1d82-7c99-4efa-b2c0-ac0d128e8964)

</details>
<details>
<summary>Older discussions</summary>

I am looking for a way to create **c# bindings for GDExtension** as stated in in the [_What's new in C# for Godot 4.0_](https://godotengine.org/article/whats-new-in-csharp-for-godot-4-0/#engine-interop-with-source-generators) article. [25 February 2023]

Given an Addon: **Addon1** with **Addon1.gdextension** with windows binary: addons/Addon1/libs/x86_64/**Addon1.dll**

![image](https://github.com/godotengine/godot-proposals/assets/49812372/6e3becd1-050c-4ef3-ba62-56b1346d60f3)

Here are the possible steps => **WIP**

![image](https://github.com/godotengine/godot-proposals/assets/49812372/679af9eb-d7f9-403d-a3b8-f33218d7953f)

### [BindingGenerator](https://github.com/antonWetzel/GDExtensionCSharp/tree/master/BindingGenerator)
> Convert Addon1.GDExtension_api.json to C#

### FYI

- [July2023:GDExtension C# with respect to Jolt Physics](https://github.com/godotengine/godot-proposals/issues/7308#issuecomment-1646280903)
- [June2023: Expose PropertyUsageFlags in GDExtension dump](https://github.com/godotengine/godot-proposals/issues/7150#issuecomment-1605362070)

- core/extension/[extension_api_dump.cpp]
- [June2023: godot-**gdextension-info**-dumper](https://github.com/RhubarbVR/godot-gdextension-info-dumper)

- [Idea from LibGodot] https://github.com/godotengine/godot/pull/72883#issuecomment-1486049029
-
</details>

**Comment by AThousandShips** [comment]
Excuse me but what are these "discussions" for? Are you just using this as a blackboard for some discussion going on elsewhere? This is a forum for discussion

Ask in the community channels, or open a proposal for this then? Also how would this be different from just using mono?

**Reply by JoNax97** [reply]
>  Also how would this be different from just using mono?

I understand that GDExtension allows for deeper access into the engine internals than what is exposed to scripting languages, specially for extending/modifying the editor itself.  Is this the case?

> open a proposal for this then?

Does this imply that C# bindings for GDExtension are not currently planned or in development? 
That would contradict [this bit](https://godotengine.org/article/whats-new-in-csharp-for-godot-4-0/#gdextension-support) of the aforementioned blog post. Could you please clarify this?

Thanks a lot

**Reply by AThousandShips** [reply]
Clarify what? These were comments a long time ago, responding to content that is no longer here, this post is used as some kind of scratch pad for some discussions elsewhere rather than a genuine discussion

**Reply by JoNax97** [reply]
@AThousandShips I understand that you are part of the core team on Godot and as such I expected you could provide more up-to-date information about the ongoing effort or plan (if any) to enable the usage of C# as a GDExtension language.

Thanks and sorry for any confusion caused.

**Reply by AThousandShips** [reply]
I'm not part of said team, but I understand

Thank you and no worries!

**Comment by GeorgeS2019** [comment]
@398utubzyt
@adamscott 
@timshannon

[NativeStructs and Pointers in C#](https://github.com/godotengine/godot/pull/81004)

I need your feedback on how best to get C# access exposed properties/methods from GDExtension written in c++

**Reply by GeorgeS2019** [reply]
@JoNax97 

https://github.com/DmitriySalnikov/godot_debug_draw_3d/issues/10#issuecomment-1712991193

**Comment by GeorgeS2019** [comment]
# Update 13th Sept 2023
Using c++

@398utubzyt
@adamscott 
@timshannon
@JoNax97

cc @DmitriySalnikov
A menu item of **Generate C# Bindings** has been added
https://github.com/DmitriySalnikov/godot_debug_draw_3d
![image](https://github.com/godotengine/godot-proposals/assets/49812372/bb0a61dc-f25b-40c7-afc7-a7e31850f117)

**Reply by DmitriySalnikov** [reply]
This is a slow version via `ClassDB` and accessing methods via strings. It is made to restore bindings that existed in previous versions of the library.
I'm still waiting for an official solution in the core of the engine itself.

**Reply by GeorgeS2019** [reply]
@DmitriySalnikov 

The feedback I received, others are inspired. 

The community needs to believe IT IS NOW Possible.  Next iteration to improve UI and optimize the performance <=== Thank you!!!

**Reply by DmitriySalnikov** [reply]
Actually, this has always been possible. These are the usual bindings via `ClassDB`. In the `Debug Draw` version for Godot 3, it worked the same as now, but the generator was written in `GDScript`, not in `C++`.
At the moment, official bindings (**not mine!**) work through native pointers, which allows you to access methods much faster.

**Reply by GeorgeS2019** [reply]
@DmitriySalnikov 
If u like, create a Godot Proposal so others could improve on what you have developed, including adding the higher performance native pointers, while improving the Usability. e.g. perhaps the menu is added under **c#: Generate Bindings: projectNameOfGDExtension**

So the proposal will become the official version

**Reply by DmitriySalnikov** [reply]
I already wrote that everything is already in the engine, except that `GDExtension` bindings need to be moved to a separate library, somehow. But most likely I'm wrong and it will take more effort.

**Reply by 398utubzyt** [reply]
I'm not sure why I keep getting mentioned here, I'm not a part of the .NET/GDExtension teams nor am I interested in GDExtension. Also, I'm not sure what your point here is exactly, but the project you linked [does not generate GDExtension bindings](https://github.com/DmitriySalnikov/godot_debug_draw_3d/blob/b7ac19661b72db443770f13b7af7b72c216e47ac/src/editor/generate_csharp_bindings.cpp#L62-L126), but instead makes C# bindings of its own classes.

**Reply by GeorgeS2019** [reply]
@398utubzyt  Good Point.  According to this version 1, The GDExtension c++ developer need to include the methods to generate the C# Bindings.  This is not an universal or generic C# Bindings Generator e.g. GDExtension <== which is future version from the community.

The main point, there is NOW a tentative way to get to C# binding FOR THE FIRST TIME
